packages/flutter/lib/src/scheduler/scheduler.dart

```diff
@@ -0,0 +1,151 @@
+import 'package:flutter/animation.dart' as animation show scheduler;
+
+/// An entry in the scheduler's priority queue.
+///
+/// Combines the task and its priority.
+class _SchedulerEntry {
+  final VoidCallback task;
+  final int priority;
+
+  const _SchedulerEntry(this.task, this.priority);
+}
+
+class Priority {
+  static const Priority idle = const Priority._(0);
+  static const Priority animation = const Priority._(100000);
+  static const Priority touch = const Priority._(200000);
+
+  /// Relative priorities are clamped by this offset.
+  ///
+  /// It is still possible to have priorities that are offset by more than this
+  /// amount by repeatedly taking relative offsets, but that's generally
+  /// discouraged.
+  static const int kMaxOffset = 10000;
+
+  const Priority._(this._value);
+
+  int get value => _value;
+  final int _value;
+
+  /// Returns a priority relative to this priority.
+  ///
+  /// A positive [offset] indicates a higher priority.
+  ///
+  /// The parameter [offset] is clamped to +/-[kMaxOffset].
+  Priority operator +(int offset) {
+    if (offset.abs() > kMaxOffset) {
+      // Clamp the input offset.
+      offset = kMaxOffset * offset.sign;
+    }
+    return new Priority._(_value + offset);
+  }
+
+  /// Returns a priority relative to this priority.
+  ///
+  /// A positive offset indicates a lower priority.
+  ///
+  /// The parameter [offset] is clamped to +/-[kMaxOffset].
+  Priority operator -(int offset) => this + (-offset);
+}
+
+/// Scheduler running tasks with specific priorities.
+///
+/// Combines the task's priority with remaining time in a frame to decide when
+/// the task should be run.
+///
+/// Tasks always run in the idle time after a frame has been committed.
+class TaskScheduler {
+  TaskScheduler._();
+
+  SchedulingStrategy schedulingStrategy = new DefaultSchedulingStrategy();
+
+  final PriorityQueue _queue = new HeapPriorityQueue<_SchedulerEntry>(
+    (_SchedulerEntry e1, _SchedulerEntry e2) {
+      // Note that we inverse the priority.
+      return -e1.priority.compareTo(e2.priority);
+    }
+  );
+
+  /// Wether this scheduler already requested to be woken up as soon as
+  /// possible.
+  bool _wakingNow = false;
+
+  /// Wether this scheduler already requested to be woken up in the next frame.
+  bool _wakingNextFrame = false;
+
+  /// Schedules the given [task] with the given [priority].
+  void schedule(VoidCallback task, Priority priority) {
+    bool isFirstTask = _queue.isEmpty;
+    _queue.add(new _SchedulerEntry(task, priority._value));
+    if (isFirstTask) _wakeNow();
+  }
+
+  /// Invoked by the system when there is time to run tasks.
+  void tick() {
+    if (_queue.isEmpty)
+      return;
+    _SchedulerEntry entry = _queue.first;
+    if (schedulingStrategy.shouldRunTaskWithPriority(entry.priority)) {
+      try {
+        (_queue.removeFirst().task)();
+      } finally {
+        if (_queue.isNotEmpty)
+          _wakeNow();
```

Comment:
That makes sense.

Another thing we could consider is spinning the event loop inside Dart.  All the incoming signals to this thread are mojo handles and we have the MojoWaitMany primitive available in Dart.  IMHO, we should see how far we can get spinning the event loop in C++ though.